cli/cmd/plugin/ui/web/tanzu-ui/src/views/management-cluster/aws/aws-mc-common/aws-orchestrator/AwsOrchestrator.service.tsx

```diff
@@ -87,4 +89,64 @@ export class AwsOrchestrator {
             setErrorObject(addErrorInfo(errorObject, e, AWS_FIELDS.NODE_TYPE));
         }
     }
+
+    static initAvailabilityZones(props: AwsOrchestratorProps) {
+        const { awsDispatch, setErrorObject, errorObject } = props;
+        DefaultOrchestrator.initResources<AWSKeyPair>({
+            resourceName: AWS_FIELDS.AVAILABILITY_ZONES,
+            dispatch: awsDispatch,
+            errorObject,
+            setErrorObject,
+            fetcher: () => AwsService.getAwsAvailabilityZones(),
+        });
+    }
+
+    static initNodeTypesForAz(props: AwsOrchestratorProps, nodeProfile: string) {
+        const { awsState, awsDispatch, setErrorObject, errorObject } = props;
+        const azList = awsState[STORE_SECTION_RESOURCES][AWS_FIELDS.AVAILABILITY_ZONES];
+        const defaultAZNameList: string[] = AwsDefaults.defaulAvailabilityZoneNameStrategy(azList, nodeProfile);
+        const selectedAzs: AvailabilityZoneInstance[] = [];
+
+        for (const az of defaultAZNameList) {
+            //  (a) store all the instance types in a segmented resource under THAT AZ
+            AwsOrchestrator.initNodeTypeForAZ(az, awsDispatch, errorObject, setErrorObject, selectedAzs, nodeProfile);
+        }
+        // (b) store the "selected" instance type in the form section.
+        awsDispatch({
+            type: INPUT_CHANGE,
+            field: AWS_FIELDS.SELECTED_AZS,
+            payload: selectedAzs,
+        } as FormAction);
+    }
+
+    static initNodeTypeForAZ(
+        az: string,
+        awsDispatch: StoreDispatch,
+        errorObject: { [fieldName: string]: any },
+        setErrorObject: (errorObject: { [fieldName: string]: any }) => void,
+        selectedAzs: AvailabilityZoneInstance[],
+        nodeProfile: string
+    ) {
+        DefaultOrchestrator.initResources<AvailabilityZoneInstance>({
+            resourceName: AWS_FIELDS.AVAILIABILITY_ZONE_NODE_TYPES,
+            segment: az,
+            dispatch: awsDispatch,
+            errorObject,
+            setErrorObject,
+            fetcher: () => {
+                return new CancelablePromise(async (res, rej) => {
+                    const nodeTypesByAZ = await AwsDefaults.createAZNodeType(az);
+                    res(nodeTypesByAZ);
+                });
+            },
+
+            fxnSelectDefault: (resources: AvailabilityZoneInstance[]) => {
+                const defaultAZ = AwsDefaults.defaulAvailabilityZoneNodeTypeStrategy(resources, nodeProfile, az);
+                if (defaultAZ !== undefined) {
+                    selectedAzs.push(defaultAZ);
```

Comment:
having this `push` here feels like a side-effect that would be nice to avoid.
Perhaps this method could return the defaultAZ and the caller could push into this array (and the caller would not need to pass in this array at all)?

Reply:
In `DefaultOrchestrator.initResources` , this 'fxnSelectDefault' will return a result which will be directly called by `dispatch` by method `saveResourceDataWithDefault`.
So if we want to use the default result outside of this `DefaultOrchestrator.initResources`, we need to pass the array as parameter, I think